File: pwa/src/sm2.js
// SM-2 spaced-repetition scheduler. Pure functions over a plain review-state
// object, mirrored 1:1 from the native version.
//
// review state shape:
//   { questionId, deckId, topic,
//     easeFactor, interval, repetitions, dueDate,
//     lastReviewedAt, totalReviews, lapses }

import { startOfDay, addDays } from "./format.js";

export const MIN_EASE = 1.3;

// SM-2 answer quality. MCQ answers are binary so only GOOD / AGAIN are used.
export const GRADE = { AGAIN: 0, HARD: 3, GOOD: 4, EASY: 5 };

export const gradeForBinary = (correct) => (correct ? GRADE.GOOD : GRADE.AGAIN);

export function newReviewState(questionId, deckId, topic, now = Date.now()) {
  return {
    questionId,
    deckId,
    topic: topic || "General",
    easeFactor: 2.5,
    interval: 0,
    repetitions: 0,
    dueDate: startOfDay(now).getTime(),
    lastReviewedAt: null,
    totalReviews: 0,
    lapses: 0,
  };
}

/**
 * Returns a NEW review-state object with SM-2 applied. Does not mutate input.
 *   - correct (q >= 3): interval 1 -> 6 -> round(interval * ease); repetitions++
 *   - incorrect (q < 3): repetitions -> 0, interval -> 1, lapses++
 *   - ease: EF += 0.1 - (5 - q)(0.08 + (5 - q)0.02), floored at 1.3
 */
export function applySM2(state, grade, now = Date.now()) {
  const q = grade;
  const s = { ...state };
  s.totalReviews += 1;

  if (q < 3) {
    s.repetitions = 0;
    s.interval = 1;
    s.lapses += 1;
  } else {
    if (s.repetitions === 0) s.interval = 1;
    else if (s.repetitions === 1) s.interval = 6;
    else s.interval = Math.round(s.interval * s.easeFactor);
    s.repetitions += 1;
  }

  const delta = 0.1 - (5 - q) * (0.08 + (5 - q) * 0.02);
  s.easeFactor = Math.max(MIN_EASE, s.easeFactor + delta);

  s.lastReviewedAt = now;
  const interval = Math.max(1, s.interval);
  s.dueDate = addDays(startOfDay(now), interval).getTime();

  return s;
}

// Non-mutating preview of the next interval in days.
export function projectedInterval({ easeFactor, interval, repetitions }, grade) {
  if (grade < 3) return 1;
  if (repetitions === 0) return 1;
  if (repetitions === 1) return 6;
  return Math.round(interval * easeFactor);
}

export function stageOf(state) {
  if (!state || state.totalReviews === 0) return "new";
  if (state.lapses > 0 && state.repetitions === 0) return "lapsed";
  if (state.repetitions < 2) return "learning";
  return "review";
}

File: DailyQCM/Session/SM2.swift
import Foundation

/// Answer quality on the SM-2 0…5 scale. MCQ answers are binary, so only two are
/// used in the app (`good` / `again`); `hard` and `easy` exist for future grading UI.
enum ReviewGrade: Int {
    case again = 0   // incorrect
    case hard  = 3
    case good  = 4   // correct
    case easy  = 5

    static func forBinary(correct: Bool) -> ReviewGrade { correct ? .good : .again }
}

/// The classic SM-2 spaced-repetition algorithm, as a pure mutation of `ReviewState`.
///
/// - Correct (quality ≥ 3): interval grows 1 → 6 → round(interval × EF); `repetitions` increments.
/// - Incorrect (quality < 3): `repetitions` resets to 0, interval to 1, a lapse is recorded.
/// - Ease factor: EF += 0.1 − (5 − q)(0.08 + (5 − q)·0.02), floored at 1.3.
enum SM2 {
    static let minEaseFactor = 1.3

    static func apply(
        to state: ReviewState,
        grade: ReviewGrade,
        on date: Date = .now,
        calendar: Calendar = .current
    ) {
        let q = grade.rawValue
        state.totalReviews += 1

        if q < 3 {
            state.repetitions = 0
            state.interval = 1
            state.lapses += 1
        } else {
            switch state.repetitions {
            case 0: state.interval = 1
            case 1: state.interval = 6
            default: state.interval = Int((Double(state.interval) * state.easeFactor).rounded())
            }
            state.repetitions += 1
        }

        let delta = 0.1 - Double(5 - q) * (0.08 + Double(5 - q) * 0.02)
        state.easeFactor = max(minEaseFactor, state.easeFactor + delta)

        state.lastReviewedAt = date
        let interval = max(1, state.interval)
        state.dueDate = calendar.date(byAdding: .day, value: interval, to: calendar.startOfDay(for: date)) ?? date
    }

    /// Non-mutating preview of the next interval — handy for tests / a "next in N days" hint.
    static func projectedInterval(
        easeFactor: Double,
        interval: Int,
        repetitions: Int,
        grade: ReviewGrade
    ) -> Int {
        guard grade.rawValue >= 3 else { return 1 }
        switch repetitions {
        case 0: return 1
        case 1: return 6
        default: return Int((Double(interval) * easeFactor).rounded())
        }
    }
}
